### PR TITLE
edgedeployment CR will have moduleName and will have type-specific settings serialized.

### PR DESCRIPTION
**Pull request by darobs**

### CODE REVIEW

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesModule.cs [hunk 1]
```diff
@@ -84,8 +99,10 @@ public bool Equals(KubernetesModule<TConfig> other)
                 this.RestartPolicy == other.RestartPolicy &&
                 this.ConfigurationInfo == other.ConfigurationInfo &&
                 this.ImagePullPolicy == other.ImagePullPolicy &&
-                this.Config is DockerConfig &&
-                this.Config as DockerConfig == other.Config as DockerConfig &&
+                this.Config is CombinedDockerConfig &&
+                this.Config as CombinedDockerConfig == other.Config as CombinedDockerConfig &&
```
Comment (kkdawkins) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesModule.cs]: Sad that you need to cast this.Config twice :(

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesModule.cs [hunk 2]
```diff
@@ -84,8 +99,10 @@ public bool Equals(KubernetesModule<TConfig> other)
...
+                this.Config is string &&
+                this.Config as string == other.Config as string &&
```
Comment (kkdawkins) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesModule.cs]: This is a really strange set of boolean conditions...

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesSpecFactory.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+namespace Microsoft.Azure.Devices.Edge.Agent.Kubernetes
+{
+    using System.Collections.Generic;
+    using Newtonsoft.Json;
+
+    public class KubernetesSpecFactory<CombinedDockerConfig> : IKubernetesSpecFactory<CombinedDockerConfig>
+    {
+        const string DockerType = "docker";
+
+        public IList<KubernetesModule<CombinedDockerConfig>> GetSpec(IList<KubernetesModule<string>> crdSpec)
```
Comment (kkdawkins) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesSpecFactory.cs]: Get Spec(s)?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesSpecFactory.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        {
+             var newList = new List<KubernetesModule<CombinedDockerConfig>>();
+             foreach (var crdModule in crdSpec)
+             {
+                 if (crdModule.Type == DockerType)
```
Comment (kkdawkins) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesSpecFactory.cs]: Is there a case where the type might come as DoCkEr? Would .tolower or ignorecase be useful here?

**Comment by kkdawkins** [comment]
The only comment that I am passionate about is the strange checks on this.config since you're casting then checking the cast, then checking that is a string. Seems wrong...

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesModule.cs (hunk 2 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesModule.cs]: Yes, let me check that more carefully - It's probably not doing what I expect.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/CrdWatcher.cs [hunk 5]
```diff
@@ -204,9 +213,10 @@ private async Task HandleEdgeDeploymentDeleted(V1ServiceList currentServices, V1
             this.currentModules = ModuleSet.Empty;
         }
 
-        private async Task UpsertDeployments(V1ServiceList currentServices, V1DeploymentList currentDeployments, EdgeDeploymentDefinition<TConfig> customObject)
+        private async Task UpsertDeployments(V1ServiceList currentServices, V1DeploymentList currentDeployments, IList<KubernetesModule<TConfig>> spec)
         {
-            var desiredModules = ModuleSet.Create(customObject.Spec.ToArray());
+            var desiredModules = ModuleSet.Create(spec.ToArray());
+            Console.WriteLine($"DESIRED MODULES:\n {desiredModules}");
```
Comment (dmolokanov) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/CrdWatcher.cs]: Shouldn't we delete all Console.WriteLine calls?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesSpecFactory.cs (hunk 4 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesSpecFactory.cs]: Ignorcase makes a lot of sense.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesModule.cs [hunk 6]
```diff
@@ -27,21 +40,23 @@ public KubernetesModule(IModule<TConfig> module)
         }
 
         [JsonConstructor]
-        public KubernetesModule(string name, string version, string type, ModuleStatus desiredStatus, RestartPolicy restartPolicy, ConfigurationInfo configurationInfo, IDictionary<string, EnvVal> env, TConfig config, ImagePullPolicy imagePullPolicy)
+        public KubernetesModule(string moduleName, string version, string type, ModuleStatus status, RestartPolicy restartPolicy, ConfigurationInfo configurationInfo, IDictionary<string, EnvVal> env, TConfig settings, ImagePullPolicy imagePullPolicy)
         {
-            this.Name = name;
+            this.Name = moduleName;
```
Comment (dmolokanov) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesModule.cs]:
```suggestion
            this.ModuleName = moduleName;
```

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesModule.cs [hunk 7]
```diff
@@ -27,21 +40,23 @@ public KubernetesModule(IModule<TConfig> module)
...
             this.Env = env?.ToImmutableDictionary() ?? ImmutableDictionary<string, EnvVal>.Empty;
-            this.Config = config;
+            this.Config = settings;
             this.ImagePullPolicy = imagePullPolicy;
         }
 
-        [JsonIgnore]
-        public string Name { get; set; }
+        [JsonProperty(PropertyName = "moduleName")]
+        public string ModuleName {get{return this.Name;}set{this.Name=value;}}
```
Comment (dmolokanov) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesModule.cs]: Should be formatted

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesModule.cs (hunk 2 above)
Comment (dmolokanov) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesModule.cs]: I agree it looks incorrect

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesSpecFactory.cs [hunk 8]
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+namespace Microsoft.Azure.Devices.Edge.Agent.Kubernetes
+{
+    using System.Collections.Generic;
+    using Newtonsoft.Json;
+
+    public class KubernetesSpecFactory<CombinedDockerConfig> : IKubernetesSpecFactory<CombinedDockerConfig>
```
Comment (dmolokanov) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesSpecFactory.cs]: What is a `spec` that you introduced here? How it is different from `EdgeDeployment`?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/CrdWatcher.cs (hunk 5 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/CrdWatcher.cs]: Yes, it should - missed that one.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesSpecFactory.cs (hunk 8 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesSpecFactory.cs]: EdgeDeployment has always had a "Spec". It's the custom part of the custom resource.

**Pull request closed without merging**